addons/membership/models/partner.py

```diff
@@ -1,253 +1,188 @@
-        list_partner = []
-        for data in data_inv:
-            list_partner.append(data.partner_id.id)
-        ids2 = list_partner
-        while ids2:
-            ids2 = res_obj.search(cr, uid, [('associate_member', 'in', ids2)], context=context)
-            list_partner += ids2
-        return list_partner
-
-    def _cron_update_membership(self, cr, uid, context=None):
-        partner_ids = self.search(cr, uid, [('membership_state', 'in', ['invoiced', 'paid'])], context=context)
-        if partner_ids:
-            self._store_set_values(cr, uid, partner_ids, ['membership_state'], context=context)
+    associate_member = fields.Many2one('res.partner', string='Associate Member',
+        help="A member with whom you want to associate your membership. "
+             "It will consider the membership state of the associated member.")
+    member_lines = fields.One2many('membership.membership_line', 'partner', string='Membership')
+    free_member = fields.Boolean(string='Free Member',
+        help="Select if you want to give free membership.")
+    membership_amount = fields.Float(string='Membership Amount', digits=(16, 2),
+        help='The price negotiated by the partner')
+    membership_state = fields.Selection(membership.STATE, compute='_compute_membership_state',
+        string='Current Membership Status', store=True,
+        help='It indicates the membership state.\n'
+             '-Non Member: A partner who has not applied for any membership.\n'
+             '-Cancelled Member: A member who has cancelled his membership.\n'
+             '-Old Member: A member whose membership date has expired.\n'
+             '-Waiting Member: A member who has applied for the membership and whose invoice is going to be created.\n'
+             '-Invoiced Member: A member whose invoice has been created.\n'
+             '-Paying member: A member who has paid the membership fee.')
+    membership_start = fields.Date(compute='_compute_membership_start',
+        string = 'Membership Start Date', store=True,
+        help="Date from which membership becomes active.")
+    membership_stop = fields.Date(compute='_compute_membership_stop',
+        string = 'Membership End Date', store=True,
+        help="Date until which membership remains active.")
+    membership_cancel = fields.Date(compute='_compute_membership_cancel',
+        string = 'Cancel Membership Date', store=True,
+        help="Date on which membership has been cancelled")
+
+    @api.depends('membership_stop', 'associate_member',
+                 'free_member', 'member_lines',
+                 'member_lines.date_to', 'member_lines.date_from',
+                 'member_lines.account_invoice_line',
+                 'member_lines.account_invoice_line.invoice_id',
+                 'member_lines.account_invoice_line.invoice_id.state',
+                 'member_lines.account_invoice_line.invoice_id.payment_ids')
+    def _compute_membership_state(self):
```

Comment:
Yep, it's a little tricky and this module is quite badly written. I'll check with him.